// types/wasm.js
const { types: { uint8, uint32le, buffer, when, array, select } } = require('binary-data')
const leb128 = require('types/leb128/signed')
const uleb128 = require('types/leb128/unsigned')

const varuint1 = uleb128(1)
const varuint7 = uleb128(7)
const varuint32 = uleb128(32)
const varint7 = leb128(7)

const Module = {
  magic: uint32le,
  version: uint32le
}

const Name = {
  length: varuint32,
  data: buffer(({ currentNode }) => currentNode.length.toNumber())
}

function ondata({ node }) {
  const length = node.length.toNumber()

  if (node.id.isZero()) {
    return length - node.name.data.length - node.name.length.byteLength()
  }

  return length
}

const value_type = varint7

const func_type = {
  form: varint7,
  param: {
    count: varuint32,
    types: array(value_type, ({ currentNode }) => currentNode.count.toNumber())
  },
  return: {
    count: varuint1,
    type: when(({ currentNode }) => currentNode.count.eqn(1), value_type)
  }
}

const SectionType = {
  count: varuint32,
  entries: array(func_type, ({ currentNode }) => currentNode.count.toNumber())
}

const SectionFunction = {
  count: varuint32,
  types: array(varuint32, ({ currentNode }) => currentNode.count.toNumber())
}

const external_kind = uint8

const export_entry = {
  length: varuint32,
  name: buffer(({ currentNode }) => currentNode.length.toNumber()),
  kind: external_kind,
  index: varuint32
}

const SectionExport = {
  count: varuint32,
  entries: array(export_entry, ({ currentNode }) => currentNode.count.toNumber())
}

const local_entry = {
  count: varuint32,
  type: value_type
}

const function_body = {
  size: varuint32,
  locals: {
    count: varuint32,
    types: array(local_entry, ({ currentNode }) => currentNode.count.toNumber())
  },
  code: buffer(onFunctionCodeSize),
  end: uint8
}

function onFunctionCodeSize({ currentNode }) {
  let size = currentNode.size.toNumber()

  size -= (currentNode.locals.count.byteLength() || 1)

  for (const entry of currentNode.locals.types) {
    size -= (entry.count.byteLength() || 1)
    size -= 1
  }

  return size - 1
}

const SectionCode = {
  count: varuint32,
  bodies: array(function_body, ({ currentNode }) => currentNode.count.toNumber())
}

const Section = {
  id: varuint7,
  length: varuint32,
  name: when(({ node }) => node.id.isZero(), Name),
  data: select(
    when(({ node }) => node.id.eqn(1), SectionType),
    when(({ node }) => node.id.eqn(3), SectionFunction),
    when(({ node }) => node.id.eqn(7), SectionExport),
    when(({ node }) => node.id.eqn(10), SectionCode),
    buffer(ondata)
  )
}

module.exports = {
  Module,
  Section,
}
